src/common/url_escape.cc

```diff
@@ -0,0 +1,72 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "url_escape.h"
+
+#include <stdexcept>
+#include <sstream>
+
+std::string url_escape(const std::string& s)
+{
+  std::string out;
+  for (auto c : s) {
+    if (c < 0x20) {
```

Comment:
for percent encoding, instead of `isprint()` (which is basically `c < 0x20` does), we need to have something like
```c
if (std::isalnum(c) || c == '-' || c == '.' || c == '_' || c == '~' || c == '/') {
  out.append(ritoa<unsigned char, 16>(c, t));
}
```